### PR TITLE
Thread pool logic and link thread lib

### PR DESCRIPTION
**Pull request by NazariiProshyn**

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -1,7 +1,29 @@
-#include "threadSafeQueue.h"
+#include "threadPool.h"
+#include <iostream>
 
+void foo1() {
```
Comment (mlohvynenko-clgx) [inline main.cpp]: remove these tests.
After this PR is merged, your next task is to use gTest to test your thread pool

File: threadPool.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+//	Date:               3/19/2021
+//	Author:             Proshyn Nazarii
+//	Description:        simple example of Thread Pool
+//	********************************************************
+
+
+#include <vector>
+#include <functional> 
+#include <thread>
+#include <queue>
```
Comment (mlohvynenko-clgx) [inline threadPool.h]: useless include

File: threadPool.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+#include <atomic>
+
+#include "threadSafeQueue.h"
+
+class ThreadPool
+{
+public:
+	//! \brief       Determines the possible number
```
Comment (mlohvynenko-clgx) [inline threadPool.h]: you constructor determinates nothing

File: threadPool.h [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	void submit(const std::function<void()> funct);
+
+private:
+	//Vector of working threads
+	std::vector<std::thread> threads;
+
+	//Indicates the possibility threads to work
+	std::atomic_bool run{true};
+
+	//Synchronized access to the task queue
```
Comment (mlohvynenko-clgx) [inline threadPool.h]: Thread-safe queue of tasks

File: threadPool.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	//! \brief                 Add function to the execution queue  
+	//! \param [in] funct      Function that is added to the execution queue
+	//! \return                   
+	void submit(const std::function<void()> funct);
+
+private:
+	//Vector of working threads
+	std::vector<std::thread> threads;
+
+	//Indicates the possibility threads to work
```
Comment (mlohvynenko-clgx) [inline threadPool.h]: I've already told you to add a space after "//" !

File: threadPool.cpp [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+		std::cout << ex.what() << std::endl;
+		throw;
+	}
+}
+
+ThreadPool::~ThreadPool()
+{
+	while (workQueue.size() > 0) {}
+	run = false;
+	for (std::vector<std::thread>::iterator iter = threads.begin(); iter < threads.end(); ++iter)
```
Comment (mlohvynenko-clgx) [inline threadPool.cpp]: why did you chose this loop?
why not for auto?

File: threadPool.cpp [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	{
+		for (size_t i = 0; i < thread_count; i++)
+		{
+			threads.push_back(
+				std::thread(&ThreadPool::workThread, this));
+		}
+	}
+	catch (const std::exception& ex)
+	{
+		run = false;
```
Comment (mlohvynenko-clgx) [inline threadPool.cpp]: what would be with items of vector that had been already created?

File: threadPool.cpp [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+		std::cout << ex.what() << std::endl;
+		throw;
+	}
+}
+
+ThreadPool::~ThreadPool()
+{
+	while (workQueue.size() > 0) {}
```
Comment (mlohvynenko-clgx) [inline threadPool.cpp]: what is the sense in this while loop?

Comment (NazariiProshyn) [reply threadPool.cpp]: we are waiting for the queue to be empty, if we do not do this we may get different types of errors or incorrect execution of the program

File: threadPool.cpp [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+		for (size_t i = 0; i < thread_count; i++)
+		{
+			threads.push_back(
+				std::thread(&ThreadPool::workThread, this));
+		}
+	}
+	catch (const std::exception& ex)
+	{
+		std::cout << ex.what() << std::endl;
+		stopvector();
```
Comment (mlohvynenko-clgx) [inline threadPool.cpp]: bad name

File: threadPool.cpp [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+		throw ex;
+	}
+}
+
+ThreadPool::~ThreadPool()
+{
+    while(workQueue.size()>0){}
```
Comment (mlohvynenko-clgx) [inline threadPool.cpp]: add comment why you need this while loop

File: threadPool.h [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    //Thread not copiable
+    ThreadPool(const ThreadPool&) = delete;
+    ThreadPool& operator=(const ThreadPool&) = delete;
+
+    /**
+     * @brief               Add function to the execution queue 
+     * 
+     * @param funct         Function that is added to the execution queue
+     */
+    void submit(const std::function<void()> funct);
```
Comment (mlohvynenko-clgx) [inline threadPool.h]: const objects can not be moved

File: threadPool.cpp [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    stopvector();
+}
+
+void ThreadPool::submit(const std::function<void()> funct)
+{
+	workQueue.push(funct);
```
Comment (mlohvynenko-clgx) [inline threadPool.cpp]: std::move(funct)

File: threadPool.cpp [hunk 13]
```diff
@@ -0,0 +1,55 @@
+#include <iostream>
+#include "threadPool.h"
+
+
+ThreadPool::ThreadPool(const size_t thread_count)
+{
+	threads.reserve(thread_count);
+	std::exception ex;
```
Comment (mlohvynenko) [inline threadPool.cpp:8]: why do you need this variable?

File: threadPool.cpp (hunk 10 above)
Comment (mlohvynenko-clgx) [reply threadPool.cpp]: spaces?

File: threadPool.cpp [hunk 14]
```diff
@@ -0,0 +1,55 @@
+#include <iostream>
```
Comment (mlohvynenko-clgx) [inline threadPool.cpp:12]: add Doxygen file descriptor

File: CMakeLists.txt [hunk 15]
```diff
@@ -3,6 +3,14 @@ project(ThreadPool)
 
 set(CMAKE_CXX_STANDART 17)
 
-add_compile_options(-Wall -Wpedantic -Werror)
+find_package( Threads )
 
-add_executable(threadPool main.cpp threadSafeQueue.h threadSafeQueue.cpp)
+if (CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
+    add_compile_options(/WX /W3 -D_WIN32_WINNT=0x0501)
```
Comment (mlohvynenko-clgx) [inline CMakeLists.txt:13]: why do you need this define "D_WIN32_WINNT"?

File: threadPool.h [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+public:
+    /**
+     * @brief                  Creates the specified number of threads  
+     * 
+     * @param thread_count     Number of threads to be created
+     */
+    ThreadPool(const size_t thread_count);
+    ~ThreadPool();
+
+    // Thread not copiable
```
Comment (mlohvynenko-clgx) [inline threadPool.h]: Is ThreadPool movable?

File: threadPool.h [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    void submit(std::function<void()> funct);
+
+private:
+
+    /**
+     * @brief       Runs functions from the queue
+     * 
+     */
+    void workThread();
```
Comment (mlohvynenko-clgx) [inline threadPool.h]: bad name

File: threadPool.h (hunk 16 above)
Comment (mlohvynenko-clgx) [inline threadPool.h]: ThreadPoool in not copiable

File: threadPool.cpp [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+#include "threadPool.h"
+
+
+ThreadPool::ThreadPool(const size_t thread_count)
+{
+	threads.reserve(thread_count);
+	std::exception ex;
```
Comment (mlohvynenko-clgx) [inline threadPool.cpp]: why do you need this variable?

File: CMakeLists.txt (hunk 15 above)
Comment (mlohvynenko-clgx) [reply CMakeLists.txt:13]: But why do you need it?
and what is "0x0501" ?

File: threadPool.cpp [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+    while( workQueue.size()>0 ) {}
+    stopThreads();
+}
+
+void ThreadPool::submit(std::function<void()> funct)
+{
+	workQueue.push(std::move(funct));
+}
+
+void ThreadPool::workingThreads()
```
Comment (mlohvynenko-clgx) [inline threadPool.cpp]: bad name

File: threadPool.h [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+     * @brief               Add function to the execution queue 
+     * 
+     * @param funct         Function that is added to the execution queue
+     */
+    void submit(std::function<void()> funct);
+
+private:
+
+    /**
+     * @brief       Runs functions from the queue
```
Comment (mlohvynenko-clgx) [inline threadPool.h]: executes tasks from the task queue

**Review by mlohvynenko-clgx — APPROVED** [review]

**Pull request merged**